### PR TITLE
Add Github Action for Industrial CI

### PR DESCRIPTION
**Pull request by schornakj**
Add CI jobs to build Yak in the various supported environments:
- Xenial with CUDA 9.2
- Xenial with CUDA 10.2
- Bionic with CUDA 9.2
- Bionic with CUDA 10.2

The Docker base images can be found [here](https://hub.docker.com/r/schornakj/yak_ci/tags) and [here](https://hub.docker.com/r/schornakj/yak-1/tags). Eventually these should be added to [the ROS-I Docker repository](https://github.com/ros-industrial/docker).

I ran into a few weird things that are partly the consequences of using ROS dependency management tools through Industrial CI to build this non-ROS package.
- ~~The `rosdep` key for OpenCV 3 is different between Kinetic and Melodic, and `libopencv-dev` installs OpenCV 2 on Xenial. I included both keys in `package.xml` and set one to be ignored for each build configuration.~~ Relaxed OpenCV verion requirement. See conversation below.
- Had to skip installing `libpcl-all-dev` via `rosdep` on Xenial, since this installs PCL 1.7 and the minimum version specified in the yak `CMakeLists.txt` is PCL 1.8. Need to do some testing to see if we actually factually need to depend on PCL 1.8 or if PCL 1.7 will work.
- Needed to pass `-DCMAKE_LIBRARY_PATH=/usr/local/cuda/lib64/stubs/` in the workspace CMake args to build successfully. This is [a known issue](https://github.com/opencv/opencv/issues/6577) when building in an environment where the CUDA library is installed but the Nvidia drivers are not installed, which is common in Docker-based CI situations.

Of course, _actually running tests_ is still a big TODO, since as a CUDA library we have specific hardware requirements. I'd be really interested to hear input on a good (and ideally free) way to set this up.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
> Of course, _actually running tests_ is still a big TODO, since as a CUDA library we have specific hardware requirements. I'd be really interested to hear input on a good (and ideally free) way to set this up.

[Self hosted runners](https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners).

@marip8 has some experience setting one up for A5 (as a test).

File: yak/package.xml [hunk 1]
```diff
@@ -11,7 +11,8 @@
   <depend>libpcl-all-dev</depend>
   <depend>eigen</depend>
   <depend>nvidia-cuda-dev</depend>
-  <depend>libopencv-dev</depend>
+  <depend>opencv3</depend> <!-- key for Xenial -->
```
Comment (gavanderhoorn) [inline yak/package.xml:14]: This is not a valid `rosdep` key: see [ros/rosdistro/rosdep/base.yaml](https://github.com/ros/rosdistro/blob/5a384b4bc7fdb0fa0a5ab211a42bcc63a5964e5d/rosdep/base.yaml) and search for `opencv3`.

For all ROS versions, `libopencv-dev` is the correct key, and should resolve to the OpenCV version distributed by the OS.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+jobs:
+  industrial_ci:
+    strategy:
+      fail-fast: false
+      matrix:
+        env:
+          - {OS_NAME: ubuntu,
+             OS_CODE_NAME: bionic,
+             ROS_DISTRO: melodic,
+             ROSDEP_SKIP_KEYS: nvidia-cuda-dev opencv3,
```
Comment (gavanderhoorn) [inline .github/workflows/main.yml:13]: You probably needed to ignore `opencv3`, as it's not a valid `rosdep` key. See my comment below.

**Comment by gavanderhoorn** [comment]
>  * The `rosdep` key for OpenCV 3 is different between Kinetic and Melodic, and `libopencv-dev` installs OpenCV 2 on Xenial. I included both keys in `package.xml` and set one to be ignored for each build configuration.

Actually, `opencv3` is not a `rosdep` key, but the name of a ROS package (this one: [wiki/opencv3](http://wiki.ros.org/opencv3)). It's the specially released version of OpenCV3 for ROS Kinetic (as Xenial comes with 2.4.9 by default).

That's why `rosdep` cannot find it for you on Bionic/Melodic: the Canonical provided version of OpenCV is used, which is a system dependency and *does* have a `rosdep` key, and the `opencv3` package is not released there for Melodic.

Using `opencv3` here is actually something to reconsider, as it would only work on a system with the OSRF repositories for ROS configured (and essentially pulls in ROS as a dependency, which IIUC should not be a dependency of plain `yak`).

**Comment by schornakj** [comment]
> Actually, `opencv3` is not a `rosdep` key, but the name of a ROS package (this one: [wiki/opencv3](http://wiki.ros.org/opencv3)). It's the specially released version of OpenCV3 for ROS Kinetic (as Xenial comes with 2.4.9 by default).
> 
> That's why `rosdep` cannot find it for you on Bionic/Melodic: the Canonical provided version of OpenCV is used, which is a system dependency and _does_ have a `rosdep` key, and the `opencv3` package is not released there for Melodic.
> 
> Using `opencv3` here is actually something to reconsider, as it would only work on a system with the OSRF repositories for ROS configured (and essentially pulls in ROS as a dependency, which IIUC should not be a dependency of plain `yak`).

I feel like I have a hard time discovering the correct `rosdep` keys for dependencies in general. This has bitten me before!

I took a look into how Yak uses OpenCV and there were only a few things that were specific to OpenCV 3. I was able to make some changes to be cross-compatible between OpenCV 2 and OpenCV 3 (at risk of overly broadening the scope of this PR):
- Replaced `cv::Ptr` with `std::shared_ptr`
- Make sure `#include` uses the older format for OpenCV module headers

This should mean that we can depend on the Canonical-provided OpenCV version, which I think is a better solution than specially-configuring the underlay environment on Xenial to include a specific OpenCV version.

**Comment by gavanderhoorn** [comment]
> which I think is a better solution than specially-configuring the underlay environment on Xenial to include a specific OpenCV version.

Agreed.

**Comment by Levi-Armstrong** [comment]
@schornakj Recommend pushing your docker files to the docker repository.

**Comment by schornakj** [comment]
> @schornakj Recommend pushing your docker files to the docker repository.

I'll add them in a PR to the [ROS Industrial Docker image repo](https://github.com/ros-industrial/docker).

**Pull request merged**